plugins/teststeps/teststeps.go

```diff
@@ -27,49 +27,46 @@ type PerTargetFunc func(cancel, pause <-chan struct{}, target *target.Target) er
 // each target. The implementation of the per-target function is responsible for
 // handling internal cancellation and pausing.
 func ForEachTarget(pluginName string, cancel, pause <-chan struct{}, ch test.TestStepChannels, f PerTargetFunc) error {
+	type tgtErr struct {
+		target *target.Target
+		err error
+	}
+	tgtErrCh := make(chan tgtErr)
 	for {
 		select {
-		case target := <-ch.In:
-			if target == nil {
+		case tgt := <-ch.In:
+			if tgt == nil {
 				// no more targets incoming
 				return nil
 			}
-			log.Debugf("%s: ForEachTarget: received target %s", pluginName, target)
-			errCh := make(chan error)
+			log.Debugf("%s: ForEachTarget: received target %s", pluginName, tgt)
 			go func() {
-				log.Debugf("%s: ForEachTarget: calling function on target %s", pluginName, target)
-				errCh <- f(cancel, pause, target)
+				log.Debugf("%s: ForEachTarget: calling function on target %s", pluginName, tgt)
+				tgtErrCh <- tgtErr{target: tgt, err: f(cancel, pause, tgt)}
 			}()
-
-			select {
-			case err := <-errCh:
-				if err != nil {
-					select {
-					case ch.Err <- cerrors.TargetError{Target: target, Err: err}:
-						log.Errorf("%s: ForEachTarget: failed to apply test step function on target %s: %v", pluginName, target, err)
-					case <-cancel:
-						log.Debugf("%s: ForEachTarget: received cancellation signal", pluginName)
-						return nil
-					case <-pause:
-						log.Debugf("%s: ForEachTarget: received pausing signal", pluginName)
-						return nil
-					}
-				} else {
-					select {
-					case ch.Out <- target:
-						log.Debugf("%s: ForEachTarget: target %s completed successfully", pluginName, target)
-					case <-cancel:
-						log.Debugf("%s: ForEachTarget: received cancellation signal", pluginName)
-						return nil
-					case <-pause:
-						log.Debugf("%s: ForEachTarget: received pausing signal", pluginName)
-						return nil
-					}
+		case te := <-tgtErrCh:
+			if te.err != nil {
+				select {
+				case ch.Err <- cerrors.TargetError{Target: te.target, Err: te.err}:
+					log.Errorf("%s: ForEachTarget: failed to apply test step function on target %s: %v", pluginName, te.target, te.err)
+				case <-cancel:
+					log.Debugf("%s: ForEachTarget: received cancellation signal", pluginName)
+					return nil
+				case <-pause:
+					log.Debugf("%s: ForEachTarget: received pausing signal", pluginName)
+					return nil
+				}
+			} else {
+				select {
+				case ch.Out <- te.target:
+					log.Debugf("%s: ForEachTarget: target %s completed successfully", pluginName, te.target)
+				case <-cancel:
+					log.Debugf("%s: ForEachTarget: received cancellation signal", pluginName)
+					return nil
+				case <-pause:
+					log.Debugf("%s: ForEachTarget: received pausing signal", pluginName)
+					return nil
 				}
-			case <-cancel:
-				return nil
-			case <-pause:
-				return nil
 			}
 		case <-cancel:
```

Comment:
Do we want to handle cancel/pause signals, while f() is running? It's necessary while we are trying to write target/error to the out channels, but here I guess plugin implementation should care about cancel/pauses signals (otherwise, why do we pass them to f() as a parameters?). Also it introduces uncertainty in behaviour: depending on  which goroutine (ForEachTarget or f()) got the signal, it could be handled in a different way, which is matter of chance.